### DISCUSSION TITLE
Loco Positioning deck on PX4

### DISCUSSION THREAD

**Discussion by afwilkin** [bitcraze/discussions#1019] [Q&A] [answered]
Hello, 

I really like the LOCO positioning system: https://www.bitcraze.io/products/loco-positioning-deck/
However, I was hoping to use it for a PX4 based (ModalAI starling) drone. It looks like the board talks over I2C. 
I was wondering what it would take to integrate the bitcraze loco hat into PX4? 

We are a team of software engineers. So, writing drivers is not a problem. Thanks!

**Comment by afwilkin** [comment]
I found this: https://www.bitcraze.io/2017/05/loco-positioning-tag/

Which is very helpful, and makes me think this can definitely be done

**Comment by krichardsson** [comment, marked as answer]
Hi,
the [Loco deck uses SPI for communication](https://www.bitcraze.io/documentation/system/platform/cf2-expansiondecks/#deck-pin-allocation).
It is probably doable to port it to another platform though it will require a bit of work. Some parts of the firmware is (pretty) platform independent and should be easy to reuse, while some other parts might require more love.

Some pointers to get started:
* [libdw1000](https://github.com/bitcraze/libdw1000), a portable driver for the DW1000 chip
* Files named `lps***.c` in [deck/drivers](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/deck/drivers/src), not very portable deck drivers for the different positioning methods but probably a good starting point.
* Files in [utils/src/tdoa](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/utils/src/tdoa), portable implementation of TDoA. There are probably dependencies to a few files outside this directory as well.

**Reply by afwilkin** [reply]
We have an onboard computer as well. I am wondering if we can pipe it through the ROS node

**Reply by afwilkin** [reply]
@krichardsson would loco positioning work correctly in a room with stationary objects. i am worried the signal will take longer if it is bouncing off those objects, and return a false distance reading. For example
![image](https://github.com/bitcraze/discussions/assets/16963678/c50d6be4-c11b-44e4-a914-1f4b860c91d6)

**Reply by afwilkin** [reply]
I know that it is recommended to keep the base stations 15cm away from walls to prevent refractions, but I am unsure how the system would handle other objects being in the way.

**Reply by knmcguire** [reply]
@afwilkin could you perhaps ask this on a different thread, especially now you selected an answer already? This will be more readable then for other users (also those that use crazyflies)

**Reply by afwilkin** [reply]
ope, apologies. Yes I will do that now
